src/main/java/org/reaktivity/nukleus/kafka/internal/stream/CachingKeyMessageDispatcher.java

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright 2016-2017 The Reaktivity Project
+ *
+ * The Reaktivity Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package org.reaktivity.nukleus.kafka.internal.stream;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.agrona.DirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.reaktivity.nukleus.kafka.internal.types.OctetsFW;
+
+public class CachingKeyMessageDispatcher extends KeyMessageDispatcher
+{
+    private final Map<UnsafeBuffer, Long> offsetsByKey = new HashMap<>();
```

Comment:
Consider storing the value as a `long[1]` instead, making it mutable and reducing GC.

Reply:
Nice idea, done